File: Java/src/PlayerAi.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class PlayerAi {
    public List<String> stones = new ArrayList<>();
    public Purse purse;

    PlayerAi(Purse purse){
        //this.stones = stones;
        this.purse = purse;
        get_rocks();
    }


    public void place_solution(Node start_node,String direction_placing,List<String> stones){
        Node actual = start_node;
        for (String num: stones){
            actual.name = String.valueOf(num);

            actual = direction_placing.equals("down") ? actual.bottom : actual.right;
        }
        remove_rocks(stones);
        get_rocks();
    }

    public void remove_rocks(List<String> rocks){
        stones.removeAll(rocks);
    }

    public void get_rocks(){
        while (stones.size() != 7){
            String stone = purse.give_stone();
            if (stone == null) break;
            else stones.add(purse.give_stone());
        }
    }

    public void change_stones(){
        List<String> new_stones = new ArrayList<>();
        for (String s : stones){
            if (!new_stones.contains(s) && !s.equals("0")){
                new_stones.add(s);
            }
            else{
                new_stones.add(purse.exchange_stone(s));
            }
        }
    }
}

File: C#/ConsoleApp1/ConsoleApp1/PlayerAi.cs
using System;
using System.Collections.Generic;

namespace ClassLibrary1
{
    public class PlayerAi
    {
        public List<string> stones = new List<string>();
        public Purse purse;

        public PlayerAi(Purse purse){
            //this.stones = stones;
            this.purse = purse;
            get_rocks();
        }


        public void place_solution(Node start_node,String direction_placing,List<String> stones){
            Node actual = start_node;
            foreach (var num in stones)
            {
                actual.name = num;
                actual = direction_placing == "down" ? actual.bottom : actual.right;
            }
            remove_rocks(stones);
            get_rocks();
        }

        public void remove_rocks(List<string> rocks){
            foreach (var rock in rocks)
            {
                stones.Remove(rock);
            }
        }

        public void get_rocks(){
            while (stones.Count != 7){
                String stone = purse.give_stone();
                if (stone == null) break;
                stones.Add(purse.give_stone());
            }
        }

        public void change_stones(){
            List<string> new_stones = new List<string>();
            foreach (var s in stones)
            {
                if (!new_stones.Contains(s) && s != "0"){
                    new_stones.Add(s);
                }
                else{
                    new_stones.Add(purse.exchange_stone(s));
                }
            }
        }
    }
}